### PR TITLE
Updated with calculation of the 21 cm forest.

### PR DESCRIPTION
**Pull request by Anikbh11**
Hey, 

Added feature to calculate optical depth for the 21 cm forest (Eq 1 in https://arxiv.org/pdf/1510.02296.pdf). However, something isn't right since this version takes forever to run (left it to run overnight and it still didnt finish running). 

Could you please look at the changes and let me know if I broke something? 

Thanks, 
Aniket

### CODE REVIEW

File: trident/absorption_spectrum/absorption_spectrum.py [hunk 1]
```diff
@@ -740,17 +749,30 @@ def _add_lines_to_spectrum(self, field_data, use_peculiar_velocity,
             n_absorbers = len(lambda_obs)
 
             # thermal broadening b parameter
-            thermal_b =  np.sqrt((2 * boltzmann_constant_cgs *
-                                  field_data['temperature']) /
-                                  line['atomic_mass'])
+            inv_temp = 1. / field_data['temperature'].d
+
+            if line['wavelength'] == 2.1e9:
+                thermal_b = np.ones(len(redshift)) * YTQuantity(self.bin_width.d,'km/s')
```
Comment (Anikbh11) [inline trident/absorption_spectrum/absorption_spectrum.py]: Setting the thermal_b to bin_width. Makes sure that absorbers inside a given bin stay inside a bin, atleast in velocity space.

File: trident/absorption_spectrum/absorption_line.py [hunk 2]
```diff
@@ -182,12 +186,20 @@ def tau_profile(lambda_0, f_value, gamma, v_doppler, column_density,
...
     global tau_factor
     if tau_factor is None:
-        tau_factor = (
-            np.sqrt(np.pi) * charge_proton_cgs ** 2 /
-            (mass_electron_cgs * speed_of_light_cgs)
-        ).in_cgs().d
+        if (lambda_0 == 2.1e9):
+            lam0 = YTQuantity(lambda_0,'angstrom')
+            gam = YTQuantity(gamma,'1/s')
+            tau_factor = (0.029842 * planck_constant_cgs * gam *
```
Comment (brittonsmith) [inline trident/absorption_spectrum/absorption_line.py:195]: Where are the Einstein coefficients and spin temperature in here?

**Review by brittonsmith — COMMENTED** [review]
Hi @Anikbh11, there are enough differences with this implementation and equation 1 of the reference that it's tough for me to pinpoint anything specific. If this is taking a very long time to run, my first guess is that the wavelength resolution is probably much too high and that it's trying to do this calculation for extremely large arrays. It's worth trying to setup a simple test with a single element where you can set the density, temperature, etc to a sensible value and verify that this is giving you what you expect. I'm unfamiliar with the expected outcome, so there's not much more I can say on this.

As for the implementation, it would be simpler to make a completely separate `tau_profile` function that accepts the arguments you need rather than making modifications in multiple places to change both the existing function and the arguments passed to it. This change might also make this simpler to debug since the changes would be confined to a single place.

**Comment by brittonsmith** [comment]
Also, I noticed you've issued this PR to my fork of the code. It may be easier for others to comment on this is you change the destination of the PR to the main repo.

**Pull request closed without merging**